external_executors/dnanexus_wes/dx-wes-lambda/dx_wes.py

```diff
@@ -0,0 +1,510 @@
+    database.
+    """
+
+    ddb = boto3.resource("dynamodb", region_name=os.environ["AWS_REGION"])
+    table = ddb.Table(os.environ["AWS_DDB_TABLE"])
+    table.put_item(
+        Item={
+            "WESWorkflowId": wes_workflow_id,
+            "EventType": "DxWDLLog",
+            "Log": log
+        }
+    )
+
+def record_dx_analysis_id(wes_workflow_id, dx_analysis_id):
+    """Record an association between a WES workflow id and a DXAnalysis id."""
+
+    ddb = boto3.resource("dynamodb", region_name=os.environ["AWS_REGION"])
+    table = ddb.Table(os.environ["AWS_DDB_TABLE"])
+
+    # If the record already exists, update it. We may have recorded a dummy
+    # value previously.
+    if lookup_dx_analysis_id(wes_workflow_id):
+        table.update_item(
+            Key={
+                "WESWorkflowId": wes_workflow_id,
+                "EventType": "DxAnalysisId"
+            },
+            UpdateExpression="set Id = :p",
+            ExpressionAttributeValues={
+                ":p": dx_analysis_id
+            }
+        )
+    else:
+        table.put_item(
+            Item={
+                "WESWorkflowId": wes_workflow_id,
+                "EventType": "DxAnalysisId",
+                "Id": dx_analysis_id
+            }
+        )
+
+
+def dx_localize_inputs(inputs_dict, dx_project_id, wes_workflow_id):
+    """Localize input files in the inputs dict to DNAnexus.
+
+    The inputs dict may contain URLs that reference remote files, and we
+    need to put those within DNAnexus before creating the workflow so the
+    workflow steps can access the files.
+
+    Also, the inputs dict might be arbitrarily nested, so we have to
+    recursively walk through it to try to find URLS to localize.
+
+    The output of this function should be the a dict with identical structure
+    as the inputs_dict, but with gs and http URLs replaced with dx URLs.
+    """
+
+    def match_url(s):
+        """Return True if s looks like a URL we can do something with."""
+        return s.startswith("http://") or \
+            s.startswith("https://") or \
+            s.startswith("gs://")
+
+    def localize_url(url):
+        """Localize a URL to a DXFile. Return a dx URL with the id of the
+        new DXFile.
+        """
+
+        # Google urls can be converted to https
+        if url.startswith("gs://"):
+            client = google.cloud.storage.client.Client.create_anonymous_client()
+
+            path_parts = url.replace("gs://", "").split("/")
+            bucket_name = path_parts[0]
+            bucket = client.bucket(bucket_name)
+            blob_name = "/".join(path_parts[1:])
+            blob = bucket.blob(blob_name)
+            public_url = blob.public_url
+        elif url.startswith("https://") or url.startswith("http://"):
+            public_url = url
+        else:
+            return url
+
+        # Create a new DXFile object and leave it open for writing
```

Comment:
It could be a good idea to either use `contextlib.closing()` or a `try: ... finally: dx_file.close()` pattern here.

Reply:
Okay, I'll add in a `try finally`. I also had to parallelize this because we're pretty time-constrained when executing this function for SS2.

Reply:
Got it. Cool.